File: app/src/main/java/lheureux/votodroid/MainActivity.java
package lheureux.votodroid;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

import lheureux.votodroid.activities.AddItemActivity;
import lheureux.votodroid.activities.QuestionAdapter;
import lheureux.votodroid.activities.ResultActivity;
import lheureux.votodroid.activities.VoteActivity;
import lheureux.votodroid.databinding.ActivityMainBinding;
import lheureux.votodroid.exceptions.InvalidQuestion;
import lheureux.votodroid.exceptions.InvalidVote;
import lheureux.votodroid.models.Question;
import lheureux.votodroid.models.Vote;
import lheureux.votodroid.services.QuestionService;
import lheureux.votodroid.services.VoteService;

public class MainActivity extends AppCompatActivity {

    private ActivityMainBinding binding;
    private QuestionAdapter questionAdapter;
    private QuestionService questionService;
    private VoteService voteService;

    private lheureux.votodroid.interfaces.onClickInterface onClickInterface;

    ActivityResultLauncher<Intent> activityLauncher = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), result -> {
        if (result.getResultCode() == RESULT_OK) {
            remplirRecycler();
        }
    });

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        binding = ActivityMainBinding.inflate(getLayoutInflater());
        View view = binding.getRoot();
        setContentView(view);

        questionService = QuestionService.getInstance(this);
        voteService = VoteService.getInstance(this);

        binding.addButton.setOnClickListener(view1 -> {
            Intent i = new Intent(MainActivity.this, AddItemActivity.class);
            activityLauncher.launch(i);
        });

        onClickInterface = ((clickedId, questionId) -> {
            switch ((int) clickedId) {
                case 0:
                    Intent i = new Intent(MainActivity.this, VoteActivity.class);
                    i.putExtra("questionId", questionId);
                    activityLauncher.launch(i);
                    break;
                case 1:
                    Intent intent = new Intent(MainActivity.this, ResultActivity.class);
                    intent.putExtra("questionId", questionId);
                    activityLauncher.launch(intent);
                    break;
                default:
                    break;
            }
        });

        initRecycler();
        remplirRecycler();

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_delete_questions:
                questionService.deleteAll();
                remplirRecycler();
                return true;
            case R.id.menu_delete_votes:
                voteService.deleteAll();
                remplirRecycler();
                return true;
            case R.id.menu_add_questions:
                if (voteService.count() > 0 && questionService.count() > 0)
                    return true;
                try {
                    questionService.insert(new Question("Que pensez-vous de Java ?"));
                    questionService.insert(new Question("Que pensez-vous de Kotlin ?"));
                    questionService.insert(new Question("Que pensez-vous de Python ?"));
                    questionService.insert(new Question("Que pensez-vous de C ?"));
                    questionService.insert(new Question("Que pensez-vous de C++ ?"));
                    questionService.insert(new Question("Que pensez-vous de C# ?"));
                    questionService.insert(new Question("Que pensez-vous de JavaScript ?"));
                    questionService.insert(new Question("Que pensez-vous de PHP ?"));
                    questionService.insert(new Question("Que pensez-vous de Ruby ?"));
                    questionService.insert(new Question("Que pensez-vous de Swift ?"));
                    questionService.insert(new Question("Que pensez-vous de Go ?"));
                    questionService.insert(new Question("Que pensez-vous de Rust ?"));
                    questionService.insert(new Question("Que pensez-vous de Dart ?"));
                    questionService.insert(new Question("Que pensez-vous de Scala ?"));
                    questionService.insert(new Question("Que pensez-vous de Groovy ?"));
                    questionService.insert(new Question("Que pensez-vous de Perl ?"));
                    questionService.insert(new Question("Que pensez-vous de Haskell ?"));
                    questionService.insert(new Question("Que pensez-vous de Lisp ?"));
                    questionService.insert(new Question("Que pensez-vous de Prolog ?"));
                    questionService.insert(new Question("Que pensez-vous de Erlang ?"));
                } catch (InvalidQuestion invalidQuestion) {
                    invalidQuestion.printStackTrace();
                }

                for (Question question : questionService.all()) {
                    int numberOfVotes = (int) (Math.random() * 50) + 50;

                    for (int i = 0; i < numberOfVotes; i++) {
                        try {
                            voteService.insert(new Vote((int) (Math.random() * 6), question.questionId, "user" + i));
                        } catch (InvalidVote e) {
                            e.printStackTrace();
                        }
                    }
                }
                remplirRecycler();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void remplirRecycler() {
        questionAdapter.list.clear();
        questionAdapter.list.addAll(questionService.all());
        questionAdapter.notifyDataSetChanged();
    }

    private void initRecycler(){
        RecyclerView recyclerView = findViewById(R.id.recyclerView);
        recyclerView.setHasFixedSize(true);

        // use a linear layout manager
        GridLayoutManager layoutManager = new GridLayoutManager(this, 1);
        recyclerView.setLayoutManager(layoutManager);

        // specify an adapter (see also next example)
        questionAdapter = new QuestionAdapter(this, new ArrayList<Question>(), onClickInterface);
        recyclerView.setAdapter(questionAdapter);
    }

}

File: app/src/main/java/com/theodorelheureux/votodroid_android/MainActivity.kt
package com.theodorelheureux.votodroid_android

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.apollographql.apollo3.api.Optional
import com.theodorelheureux.votodroid_android.activities.*
import com.theodorelheureux.votodroid_android.databinding.ActivityMainBinding
import com.theodorelheureux.votodroid_android.graphql.GQLClient
import com.theodorelheureux.votodroid_android.interfaces.onClickInterface
import com.theodorelheureux.votodroid_android.services.UserService
import kotlinx.coroutines.*


class MainActivity : AppCompatActivity() {
    private var binding: ActivityMainBinding? = null
    private var questionAdapter: QuestionAdapter? = null
    private var onClickInterface: onClickInterface = object : onClickInterface {
        override fun setClick(clickedId: Int, questionId: String, questionText: String) {
            when (clickedId) {
                0 -> {
                    val intent = Intent(this@MainActivity, VoteActivity::class.java)
                    intent.putExtra("questionId", questionId)
                    intent.putExtra("questionText", questionText)
                    startActivity(intent)
                }
                1 -> {
                    val intent = Intent(this@MainActivity, ResultActivity::class.java)
                    intent.putExtra("questionId", questionId)
                    intent.putExtra("questionText", questionText)
                    startActivity(intent)
                }
                else -> {
                    return
                }
            }
        }
    }

    private var cursor: String? = null
    private var isLoading = false
    private var hasMore = false
    private val pageSize = 10

    var activityLauncher =
        registerForActivityResult(StartActivityForResult()) { result: ActivityResult ->
            if (result.resultCode == RESULT_OK) {
                remplirRecycler()
            }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(
            layoutInflater
        )
        val view: View = binding!!.root
        val swipeRefresh = binding!!.swiperefresh
        setContentView(view)

        val user = runBlocking { UserService.getUser() }

        if (user == null) {
            val intent = Intent(this, LoginActivity::class.java)
            activityLauncher.launch(intent)
        }

        binding!!.addButton.setOnClickListener { view1 ->
            val i = Intent(this@MainActivity, AddItemActivity::class.java)
            activityLauncher.launch(i)
        }
        binding!!.mainProgressBar.visibility = View.GONE
        initRecycler()
        remplirRecycler()
        initScrollListener()

        swipeRefresh.setOnRefreshListener {
            remplirRecycler()
            swipeRefresh.isRefreshing = false
        }

    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.main_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_logout -> {
                runBlocking { UserService.logout() }
                val intent = Intent(this, LoginActivity::class.java)
                activityLauncher.launch(intent)
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun remplirRecycler() {
        questionAdapter!!.list.clear()
        cursor = null
        hasMore = false
        isLoading = false
        loadMore()
    }

    private fun initRecycler() {
        val recyclerView = findViewById<RecyclerView>(R.id.recyclerView)
        recyclerView.setHasFixedSize(true)

        // use a linear layout manager
        val layoutManager = GridLayoutManager(this, 1)
        recyclerView.layoutManager = layoutManager

        // specify an adapter (see also next example)
        questionAdapter = QuestionAdapter(ArrayList(), onClickInterface)
        recyclerView.adapter = questionAdapter
    }

    private fun initScrollListener() {
        binding!!.recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                val linearLayoutManager = recyclerView.layoutManager as LinearLayoutManager?

                if (!isLoading && hasMore) {
                    if (linearLayoutManager != null && linearLayoutManager.findLastCompletelyVisibleItemPosition() == questionAdapter!!.list.size - 1) {
                        isLoading = true
                        binding!!.mainProgressBar.visibility = View.VISIBLE
                        loadMore()
                    }
                }
            }
        })
    }

    private fun loadMore() {
        CoroutineScope(Dispatchers.IO).launch {
            var query =
                GQLClient.instance.query(GetPaginatedQuery(pageSize, Optional.present(cursor)))
                    .execute()

            if (query.data?.questions?.getPaginated?.questions == null) {
                return@launch
            }

            withContext(Dispatchers.Main) {
                var questions = query.data!!.questions.getPaginated.questions!!

                if (questions.size == pageSize) {
                    hasMore = true
                    cursor = questions[questions.size - 1].id.toString()
                    questions = questions.subList(0, questions.size - 2)
                } else {
                    hasMore = false
                }

                questionAdapter!!.list.addAll(questions)
                questionAdapter!!.notifyDataSetChanged()

                binding!!.mainProgressBar.visibility = View.GONE
                isLoading = false
            }
        }
    }
}